Firestore/Example/Tests/Integration/FSTStreamTests.m

```diff
@@ -0,0 +1,261 @@
+             snapshotVersion:(FSTSnapshotVersion *)snapshotVersion {
+  [_states addObject:@"watchStreamDidChange"];
+  [_expectation fulfill];
+  _expectation = nil;
+}
+
+- (void)writeStreamDidReceiveResponseWithVersion:(FSTSnapshotVersion *)commitVersion
+                                 mutationResults:(NSArray<FSTMutationResult *> *)results {
+  [_states addObject:@"writeStreamDidReceiveResponseWithVersion"];
+  [_expectation fulfill];
+  _expectation = nil;
+}
+
+- (void)fulfillOnCallback:(XCTestExpectation *)expectation {
+  FSTAssert(_expectation == nil, @"Previous expectation still active");
+  _expectation = expectation;
+}
+
+@end
+
+@interface FSTStreamTests : XCTestCase
+
+@end
+
+@implementation FSTStreamTests {
+  dispatch_queue_t _testQueue;
+  FSTDatabaseInfo *_databaseInfo;
+  FSTEmptyCredentialsProvider *_credentials;
+  FSTStreamStatusDelegate *_delegate;
+  FSTDispatchQueue *_workerDispatchQueue;
+
+  /** Single mutation to send to the write stream. */
+  NSArray<FSTMutation *> *_mutations;
+}
+
+- (void)setUp {
+  [super setUp];
+
+  _mutations = @[ FSTTestSetMutation(@"foo/bar", @{}) ];
+
+  FIRFirestoreSettings *settings = [FSTIntegrationTestCase settings];
+  FSTDatabaseID *databaseID =
+      [FSTDatabaseID databaseIDWithProject:[FSTIntegrationTestCase projectID]
+                                  database:kDefaultDatabaseID];
+
+  _databaseInfo = [FSTDatabaseInfo databaseInfoWithDatabaseID:databaseID
+                                               persistenceKey:@"test-key"
+                                                         host:settings.host
+                                                   sslEnabled:settings.sslEnabled];
+  _testQueue = dispatch_queue_create("FSTStreamTestWorkerQueue", DISPATCH_QUEUE_SERIAL);
+  _workerDispatchQueue = [FSTDispatchQueue queueWith:_testQueue];
+  _credentials = [[FSTEmptyCredentialsProvider alloc] init];
+}
+
+- (void)tearDown {
+  [super tearDown];
+}
+
+- (FSTWriteStream *)setUpWriteStream {
+  FSTDatastore *datastore = [[FSTDatastore alloc] initWithDatabaseInfo:_databaseInfo
+                                                   workerDispatchQueue:_workerDispatchQueue
+                                                           credentials:_credentials];
+
+  _delegate = [FSTStreamStatusDelegate new];
+  return [datastore createWriteStreamWithDelegate:_delegate];
+}
+
+- (FSTWatchStream *)setUpWatchStream {
+  FSTDatastore *datastore = [[FSTDatastore alloc] initWithDatabaseInfo:_databaseInfo
+                                                   workerDispatchQueue:_workerDispatchQueue
+                                                           credentials:_credentials];
+
+  _delegate = [FSTStreamStatusDelegate new];
+  return [datastore createWatchStreamWithDelegate:_delegate];
+}
+
+- (void)verifyDelegate:(NSArray<NSString *> *)expectedStates {
+  // Drain queue
+  dispatch_sync(_testQueue, ^{
+                });
+
+  XCTAssertEqualObjects(_delegate.states, expectedStates);
```

Comment:
Shouldn't this also remove the states accumulated in the delegate so that subsequent assertions only need to assert about the new states?

Reply:
Hm, we can do this here, but it's not technically needed since `delegate` is scoped to the test.

Reply:
`delegate` is scoped to the test but not clearing this means that you can't reasonably make more than one assertion. I'm also fine just leaving this alone until we need to do that, but just wanted the intent of my comment to be clear.